src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -248,9 +248,12 @@ private IList<IModelValidator> GetValidators(IModelValidatorProvider provider, M
                 }
             }
 
-            if (isValid)
+            if (modelState.ContainsKey(modelKey))
             {
-                validationContext.ModelValidationContext.ModelState.MarkFieldValid(modelKey);
+                if (isValid)
+                {
+                    validationContext.ModelValidationContext.ModelState.MarkFieldValid(modelKey);
+                }
```

Comment:
>Is the issue here that MarkFieldValid creates another entry if it doesn't exist? It would be good to leave a comment to that effect.

Yes. Specifically we want to only add an entry if 
a. there is an error.
b. there is a value set by a model binder and that needs to be set to valid, invalid or skipped.

> nit: also can combine the two conditions to reduce nesting and avoid the lookup if it's not valid.

yeah, missed that.